Add PDF Viewer

tobiasdiez opened the pull request:
<!-- describe the changes you have made here: what, why, ... -->

Since @lynyus wants to implement a fulltext pdf search 😛, we need a proper way to display the results. In this PR a very plain PDF viewer is implemented. Most  #PDFs are displayed correctly, maybe with a few small glitches here and there. Basic features like scrolling, changing pages and zooming work. In addition, while being in "live mode" the linked document of the currently selected BibEntry is shown - this is very handy in a two-display setup since one can have JabRef's maintable on one screen and the always-current PDF on the second.

![untitled](https://cloud.githubusercontent.com/assets/5037600/24460654/3d87a47c-149f-11e7-98ba-7ccabf5917d6.png)

Feedback welcome! (I will update the language files before merge.)

- Similar project: https://github.com/IDRSolutions/maven-OpenViewerFX-src
- Extended example: https://github.com/james-d/PdfViewer/blob/master/src/pdfviewer/PdfViewer.fxml and https://github.com/james-d/PdfViewer/blob/master/src/pdfviewer/Controller.java

----

- [x] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [x] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

stefan-kolb commented:
Code looks ok to me 👍 Need to try it out, still.

Siedlerchr commented on src/main/java/org/jabref/gui/documentviewer/DocumentViewerController.java:
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (C) 2003-2016 JabRef contributors.
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
```
The header still days GPL

tobiasdiez replied:
Good catch! Some of the code was written in an era, when such copyright notices were still included 😄

stefan-kolb commented:
Ok, tested it a bit. Some thoughts:
- live view is nice 👍 
- functionality currently quite hidden via `view - show document viewer`. Probably need a magnifier glass somewhere. Not sure where tho.
- PDF rendering is a bit slow.
- Also, it does not render backgrounds correctly and also the casing of some texts seem to be odd
![image](https://cloud.githubusercontent.com/assets/2141507/24503167/331d73d8-1551-11e7-9ae3-a13ee522c316.png)
![image](https://cloud.githubusercontent.com/assets/2141507/24503174/3a50b9f8-1551-11e7-8d85-a7ea0ab0d13f.png)
![image](https://cloud.githubusercontent.com/assets/2141507/24503215/5f2fdf7e-1551-11e7-8679-0bd5a550be5b.png)
![image](https://cloud.githubusercontent.com/assets/2141507/24503221/6662719e-1551-11e7-8f86-f96fc6ddd888.png)
- Highlighting does not display correctly
![image](https://cloud.githubusercontent.com/assets/2141507/24503246/8d55f92e-1551-11e7-8341-7ef39af77d77.png)
- Comments cannot be seen
- I can no longer navigate through the main panel with UP/DOWN buttons

__We already had such a viewer before and removed it__. Isn't it possible to do this natively with Acrobat or so? I doubt we will ever get a really good rendering by ourselves.

tobiasdiez commented:
Thanks for the feedback. I also noticed some of these rendering issues (and, in addition, some math characters are displaced or not shown at all). I did some investigation and it seems that most of these problems are fixed/improved upon with PDFBox 2.0. But the update to 2.0 is still not possible for us https://github.com/JabRef/jabref/pull/1096. The only other decent PDF rendering engine I found for java is itext, but it is licensed under the incompatible AGPL (this is also the reason why #2474 is blocked at the moment). 

We already had a PDF viewer? I was only aware of the PDF thumbnail, which was totally buged.
The idea behind this PR is not to implement a full-fledged PDF viewer and editor, but only provide a reasonable good minimal viewer. Its main purpose should be to let the user quickly decide whether he has picked the right document for further reading. In addition, I plan to show the preview also in the import dialog to give more context there. Some further plans are integration with and enrichment of the core bibliographic features of JabRef: pdf fulltext search, visualization of cites / cited by relationships, automatic import of referenced literature.
Thus the idea is kind of similar to the PDF previewer in firefox: provide a means for a quick inline PDF preview for convenience but don't try to write yet another PDF editor. 

I will have a look why the navigation in the maintable with arrow keys does not work anymore.

lenhard commented:
And one more thing that just came to my mind: What are the licenses of the libraries you are including here?

tobiasdiez commented on src/main/java/org/jabref/model/entry/ParsedFileField.java:
```diff
@@ -1,7 +1,14 @@
 import java.util.Objects;
+import java.util.Optional;
+
+import org.jabref.logic.util.io.FileUtil;
```
May I also put `FileUtil` in `model` since it is our abstraction of the `File` object? Reason: I really want the `ParsedFileField` to have a method that returns the path to the linked file. There are around 15 places where `expandFilename` is used upon `ParsedFileField::getLink`.

lenhard replied:
The thing is, `FileUtil` has more dependencies to logic classes, which would then also need to move to model. At some point, the distinction between the two packages would be completely lost.

Can't you just add a method `public static Path getPath(ParsedFileField fileField)` to `FileUtil` instead? Not much better than the current version, I know, but at least a little. Alternatively, extract the methods you need, and just those, into a new `FileUtil` class in model (maybe with a better name). It wouldn't hurt to break up these util classes into more meaningful smaller units. They all devolve into god classes...

tobiasdiez replied:
I moved the method to FileUtil as you proposed and will open a new PR with the refactoring since it turned out to be a bit more extensive than I have expected.

tobiasdiez commented:
@lenhard Thanks for the additional feedback. 
I changed the log output of pdfbox to errors, so all the warnings are hidden now. The "-fx-background-color" messages seem to be a bug in JavaFX at least it referenced a code point in JavaFX's own stylesheet `modena.css`. 

Moreover, I added the licenses: Apache 2.0 and BSD 2, so this shouldn't be a problem.
